crypto/crmf/crmf_lib.c

```diff
@@ -612,102 +646,297 @@ int OSSL_CRMF_CERTTEMPLATE_fill(OSSL_CRMF_CERTTEMPLATE *tmpl,
+        goto end;
+    }
+
+    if (pkey_bio == NULL) {
+        ERR_raise(ERR_LIB_CRMF, CRMF_R_ERROR_VERIFYING_ENCRYPTEDKEY);
+        goto end;
+    }
+
+    /* unpack AsymmetricKeyPackage */
+    if ((ret = d2i_PrivateKey_ex_bio(pkey_bio, NULL, libctx, propq)) == NULL)
+        ERR_raise(ERR_LIB_CRMF, CRMF_R_ERROR_DECODING_ENCRYPTEDKEY);
+
+ end:
+    CMS_SignedData_free(sd);
+    BIO_free(bio);
+    BIO_free(pkey_bio);
+    return ret;
+#else
+    /* prevent warning on unused parameters: */
+    ((void)ts, (void)extra, (void)cert, (void)secret);
+    ERR_raise(ERR_LIB_CRMF, CRMF_R_CMS_NOT_SUPPORTED);
+    return NULL;
+#endif /* OPENSSL_NO_CMS */
+}
+
+unsigned char
+*OSSL_CRMF_ENCRYPTEDVALUE_decrypt(const OSSL_CRMF_ENCRYPTEDVALUE *enc,
+                                  OSSL_LIB_CTX *libctx, const char *propq,
+                                  EVP_PKEY *pkey, int *outlen)
 {
-    X509 *cert = NULL; /* decrypted certificate */
     EVP_CIPHER_CTX *evp_ctx = NULL; /* context for symmetric encryption */
     unsigned char *ek = NULL; /* decrypted symmetric encryption key */
     size_t eksize = 0; /* size of decrypted symmetric encryption key */
-    EVP_CIPHER *cipher = NULL; /* used cipher */
+    const EVP_CIPHER *cipher = NULL; /* used cipher */
     int cikeysize = 0; /* key size from cipher */
     unsigned char *iv = NULL; /* initial vector for symmetric encryption */
-    unsigned char *outbuf = NULL; /* decryption output buffer */
-    const unsigned char *p = NULL; /* needed for decoding ASN1 */
-    int n, outlen = 0;
+    unsigned char *out = NULL; /* decryption output buffer */
+    int symmAlg = 0; /* NIDs for symmetric algorithm */
+    int n, ret = 0;
     EVP_PKEY_CTX *pkctx = NULL; /* private key context */
-    char name[OSSL_MAX_NAME_SIZE];
 
-    if (ecert == NULL || ecert->symmAlg == NULL || ecert->encSymmKey == NULL
-            || ecert->encValue == NULL || pkey == NULL) {
+    if (outlen == NULL) {
+        ERR_raise(ERR_LIB_CRMF, CRMF_R_NULL_ARGUMENT);
+        return NULL;
+    }
+    *outlen = 0;
+    if (enc == NULL || enc->symmAlg == NULL || enc->encSymmKey == NULL
+            || enc->encValue == NULL || pkey == NULL) {
         ERR_raise(ERR_LIB_CRMF, CRMF_R_NULL_ARGUMENT);
         return NULL;
     }
 
-    /* select symmetric cipher based on algorithm given in message */
-    OBJ_obj2txt(name, sizeof(name), ecert->symmAlg->algorithm, 0);
-
-    (void)ERR_set_mark();
-    cipher = EVP_CIPHER_fetch(NULL, name, NULL);
-
-    if (cipher == NULL)
-        cipher = (EVP_CIPHER *)EVP_get_cipherbyname(name);
+    if ((symmAlg = OBJ_obj2nid(enc->symmAlg->algorithm)) == 0) {
+        ERR_raise(ERR_LIB_CRMF, CRMF_R_UNSUPPORTED_CIPHER);
+        return NULL;
+    }
 
-    if (cipher == NULL) {
-        (void)ERR_clear_last_mark();
+    /* select symmetric cipher based on algorithm given in message */
+    if ((cipher = EVP_get_cipherbynid(symmAlg)) == NULL) {
```

Comment:
This is wrong, please revert this change.

Reply:
I reverted and restructured it little. Please check.